### PR TITLE
B-142: several nics same vnet

### LINKED ISSUE #142: [Virtual machine] can't attach several NICs to the same network
For now, due to the exclusive use of the `network_id` to compare NICs in some code parts, it's not possible to attach several NICs to the same network.

The equivalent issue for disks, i.e. not possible to have several NIC using the same image: #125

### PR DESCRIPTION
**Pull request by treywelsh**
This PR also try to fix the case when using some `template_id` with predefined NIC

### CODE REVIEW

File: opennebula/helpers_vm.go [hunk 1]
```diff
@@ -137,45 +138,93 @@ func vmDiskResize(vmc *goca.VMController, timeout, diskID, newsize int) error {
...
 
 	networkID, err := nicTpl.GetI(shared.NetworkID)
 	if err != nil {
-		return fmt.Errorf("NIC template doesn't have a network ID")
+		return -1, fmt.Errorf("NIC template doesn't have a network ID")
 	}
 
 	log.Printf("[DEBUG] Attach NIC to network (ID:%d)", networkID)
 
+	// Retrieve disk list
```
Comment (jaypif) [inline opennebula/helpers_vm.go]: Typo ;)

File: opennebula/helpers_vm.go [hunk 2]
```diff
@@ -137,45 +138,93 @@ func vmDiskResize(vmc *goca.VMController, timeout, diskID, newsize int) error {
...
 	}
 
-	// If NIC not attached, retrieve error message
-	vmerr, _ := vm.UserTemplate.Get(vmk.Error)
+	var attachedNIC *shared.NIC
+
+	switch len(oldNICs) {
+	case 0:
+
+		// If disk not attached, retrieve error message
```
Comment (jaypif) [inline opennebula/helpers_vm.go]: same typo here

File: opennebula/helpers_vm.go [hunk 3]
```diff
@@ -137,45 +138,93 @@ func vmDiskResize(vmc *goca.VMController, timeout, diskID, newsize int) error {
...
+		vmerr, _ := vm.UserTemplate.Get(vmk.Error)
+
+		return -1, fmt.Errorf("network %d: %s", networkID, vmerr)
+
+	case 1:
+		attachedNIC = &oldNICs[0]
+	default:
+	loop:
+		for i, disk := range oldNICs {
```
Comment (jaypif) [inline opennebula/helpers_vm.go]: replace `disk` by `nic`

**Review by jaypif — APPROVED** [review]

**Comment by treywelsh** [comment]
Only a small typo fixed

File: opennebula/helpers_vm.go [hunk 4]
```diff
@@ -137,45 +138,93 @@ func vmDiskResize(vmc *goca.VMController, timeout, diskID, newsize int) error {
...
 		}
+
+		oldNICs = append(oldNICs, nic)
 	}
 
-	// If NIC not attached, retrieve error message
-	vmerr, _ := vm.UserTemplate.Get(vmk.Error)
+	var attachedNIC *shared.NIC
+
+	switch len(oldNICs) {
```
Comment (jaypif) [inline opennebula/helpers_vm.go]: I don't understand the usage of switch here. 
Why not testing len to 0 and in other case doing the loop (event for len=1) ?

File: opennebula/helpers_vm.go [hunk 5]
```diff
@@ -137,45 +138,93 @@ func vmDiskResize(vmc *goca.VMController, timeout, diskID, newsize int) error {
...
+	case 0:
+
+		// If nic not attached, retrieve error message
+		vmerr, _ := vm.UserTemplate.Get(vmk.Error)
+
+		return -1, fmt.Errorf("network %d: %s", networkID, vmerr)
+
+	case 1:
```
Comment (jaypif) [inline opennebula/helpers_vm.go]: this case means no change right ?

File: opennebula/resource_opennebula_virtual_machine_test.go [hunk 6]
```diff
@@ -955,6 +1012,10 @@ var testAccVirtualMachineTemplateConfigNICIPUpdate = testNICVNetResources + `
 		customer = "test"
 	  }
 
+	  nic {
```
Comment (jaypif) [inline opennebula/resource_opennebula_virtual_machine_test.go:1081]: it seems there is twice same test

File: opennebula/helpers_vm.go (hunk 4 above)
Comment (treywelsh) [reply opennebula/helpers_vm.go:191]: Among others, It seems that I used some misleading names in this function.
`set` keeps NICs presents before the update and I probably should rename it `refNICs` 
`oldNICs` contains the NIC that are not in `refNICSet`. I should rename it `updatedNICs`

Indeed, I'll try to simplify this function.

Comment (treywelsh) [reply opennebula/helpers_vm.go]: It seems I also used misleading names for `set` and `oldNICs`.

Indeed, I'll update this for something simpler and more reliable

File: opennebula/helpers_vm.go (hunk 5 above)
Comment (treywelsh) [reply opennebula/helpers_vm.go]: This means: only one NIC has been updated, It's probably the one I'm looking for so it's the normal case.
But I realize that I didn't check if it's really the NIC I'm looking for...
I'll update this

File: opennebula/resource_opennebula_virtual_machine_test.go (hunk 6 above)
Comment (treywelsh) [reply opennebula/resource_opennebula_virtual_machine_test.go:1081]: Here, there is two NICs, I updated one of them and l left the other unmodified.
Sometimes I try to have several elements in a TypeList, and I add unmodified parts, it's a part of the test.

**Review by jaypif — APPROVED** [review]

**Pull request merged**